core/datatrak.c

```diff
@@ -123,6 +123,23 @@ static int dtrak_prepare_data(int model, device_data_t *dev_data)
 	return g_models[i].libdc_num;
 }
 
+/*
+ * Return a default name for a tank based on it's size.
+ * Just get the first in the user's list for given size.
+ * Reaching the end of the list means there is no tank of this size.
+ */
+static char *cyl_type_by_size(int size)
+{
+	struct tank_info_t *ti = tank_info;
+
+	while (ti->ml != size && ti < tank_info + MAX_TANK_INFO)
+		ti++;
+	if (ti == tank_info + MAX_TANK_INFO)
+		return "";
```

Comment:
The memory management here is messed up. Below you return a copy of a string, here a static string - that can't be right. This should give a fat compiler warning (conversion of `const char *` to `char *`).

PS: Yes, I am aware that this condition is nearly impossible to reach. However, that makes it even more severe - it may lead to very obscure crashes.

Reply:
Reading the code below - this particular line is probably correct, but you should loose the copy_string below. The cylinder appears to be cloned anyway. Or you copy the string this function and then use `add_cylinder` instead of `add_cloned_cylinder`. Of course that should be commented. The joys of C-style memory management.

Reply:
Yes, using add_cloned_cylinder() works well.  I don't think this should be moved to add_cylinder() as we are not freeing the original string anyway.
On the previous comment, I'm not getting any warning because of "" const char * being returned.